lucene/analysis/common/src/java/org/apache/lucene/analysis/pattern/PatternTypingFilterFactory.java

```diff
@@ -0,0 +1,120 @@
+ *   &lt;analyzer&gt;
+ *     &lt;tokenizer class="solr.WhitespaceTokenizerFactory"/&gt;
+ *     &lt;filter class="com.example.PatternTypingFilter" patternFile="patterns.txt"/&gt;
+ *     &lt;filter class="solr.TokenAnalyzerFilter" asType="text_en" preserveType="true"/&gt;
+ *     &lt;filter class="solr.TypeAsSynonymFilterFactory" prefix="__TAS__"
+ *               ignore="word,&amp;lt;ALPHANUM&amp;gt;,&amp;lt;NUM&amp;gt;,&amp;lt;SOUTHEAST_ASIAN&amp;gt;,&amp;lt;IDEOGRAPHIC&amp;gt;,&amp;lt;HIRAGANA&amp;gt;,&amp;lt;KATAKANA&amp;gt;,&amp;lt;HANGUL&amp;gt;,&amp;lt;EMOJI&amp;gt;"/&gt;
+ *   &lt;/analyzer&gt;
+ * &lt;/fieldType&gt;</pre>
+ * <p>
+ * Note that a configuration such as above may interfere with multi-word synonyms. The patterns file has the format:
+ * <pre>
+ * (flags) (pattern) ::: (replacement)
+ * </pre>
+ * Therefore to set the first 2 flag bits on the original token matching 401k or 401(k) and adding a type of
+ * 'legal2_401_k' whenever either one is encountered one would use:
+ * <pre>
+ * 3 (\d+)\(?([a-z])\)? ::: legal2_$1_$2
+ * </pre>
+ * Note that the number indicating the flag bits to set must not have leading spaces and be followed by a single
+ * space, and must be 0 if no flags should be set. The flags number should not contain commas or a decimal point.
+ * Lines for which the first character is <code>#</code> will be ignored as comments.  Does not support producing
+ * a synonym textually identical to the original term.
+ *
+ * @lucene.spi {@value #NAME}
+ * @since 8.8
+ */
+public class PatternTypingFilterFactory extends TokenFilterFactory implements ResourceLoaderAware {
+
+  /**
+   * SPI name
+   */
+  public static final String NAME = "patternTyping";
+
+  private final String patternFile;
+  // todo, perhaps this could be turned into an FST (wanted: regex FST with capturing groups in java)
+  private final LinkedHashMap<Pattern, String> patterns = new LinkedHashMap<>();
+  private final Map<Pattern, Integer> flags = new HashMap<>();
+
+  /**
+   * Creates a new KeepWordFilterFactory
+   */
+  public PatternTypingFilterFactory(Map<String, String> args) {
+    super(args);
+    patternFile = require(args, "patternFile");
+    if (!args.isEmpty()) {
+      throw new IllegalArgumentException("Unknown parameters: " + args);
+    }
+  }
+
+  /** Default ctor for compatibility with SPI */
+  public PatternTypingFilterFactory() {
+    throw defaultCtorException();
+  }
+
+  @Override
+  public void inform(ResourceLoader loader) throws IOException {
+    List<String> lines = getLines(loader, patternFile);
+    // format: # regex ::: typename[_$1[_$2 ...]]    (technically _$1 does not need the '_' but it usually makes sense)
+    // eg: 2 (\d+\(?([a-z])\)?\(?(\d+)\)? ::: legal3_$1_$2_3
+    // which yields legal3_501_c_3 for 501(c)(3) or 501c3 and sets the second lowest bit in flags
+    for (String line : lines) {
+      if (line.trim().startsWith("#")) {
```

Comment:
This part seems to be checked in WordlistLoader.getLines!
https://github.com/apache/lucene-solr/blob/f7be9e83e43a549aa92ca6dc26972325f418100a/lucene/core/src/java/org/apache/lucene/analysis/AbstractAnalysisFactory.java#L280-L282

https://github.com/apache/lucene-solr/blob/f7be9e83e43a549aa92ca6dc26972325f418100a/lucene/core/src/java/org/apache/lucene/analysis/WordlistLoader.java#L217-L227